plugin.py

```diff
@@ -0,0 +1,247 @@
+            session.window.show_input_panel(
+                "Follow the in-browser instructions and paste here the final URL after the indirection:", "",
+                lambda r: self._send_response_async(r), None, lambda: self._send_response_async(None))
+
+    def _on_error_async(self, error: Any) -> None:
+        self.error_message("Failed to start login process")
+
+    def _send_response_async(self, input: Optional[str]) -> None:
+        if input:
+            res = urlparse(input)
+            if res.scheme != 'vscode':
+                self.error_message("Unexpected authorization URL schema. Hint: the inputted authorization URL should start with 'vscode://'")
+                return
+            if self.weaksession:
+                self._send_authorization_url(input, self.weaksession)
+
+
+class LspGrammarlyExecuteLoginThroughThirdPartyCommand(LspGrammarlyExecuteLoginCommand):
+    """
+    We piggyback on VSCode client's functionality and redirect utility.
+
+    To avoid the forbidden redirect to localhost, we use the same
+    netlify function as the VSCode plugin to redirect to the localhost.
+    **Without any guarantee**, the VSCode netlify functions should be
+    running the code from here:
+    https://github.com/znck/grammarly/blob/main/redirect/functions/redirect.js
+    """
+    redirect_uri = 'https://vscode-extension-grammarly.netlify.app/.netlify/functions/redirect'
+    localhost_qs = '?vscode-scheme=vscode&vscode-authority=znck.grammarly&vscode-path=/auth/callback'
+
+    def run(self, edit: sublime.Edit) -> None:
+        session = self.session_by_name()
+        if not session:
+            return
+
+        weaksession = weakref.ref(session)
+        weaklogincmd = weakref.ref(self)
+
+        class HandleAuthResponse(BaseHTTPRequestHandler):
+            def do_GET(self):
+                self.send_response(200)
+                self.send_header('Content-type', 'text/html')
+                self.end_headers()
+                self.wfile.write(r"""
+                    <html>
+                        <link rel="icon" href="data:;base64,iVBORw0KGgo=">
+                        <head><title>LSP-Grammarly</title></head>
+                        <body onload="window.close()">
+                            LSP-Grammarly: You can close this tab
+                        </body>
+                    </html>""".encode('utf-8'))
+                logincmd = weaklogincmd()
+                if not logincmd:
+                    return
+                if not self.path:
+                    return
+                qkeys = parse_qs(self.path)
+                code = qkeys["code"]
+                if not code:
+                    return
+                input = LspGrammarlyExecuteLoginThroughThirdPartyCommand.internal_redirect_uri + '?code=' + code[0]
+                logincmd._send_authorization_url(input, weaksession)
+
+        server = HTTPServer(('localhost', 0), HandleAuthResponse)
+        server.timeout = 5 * 60  # 5 min
```

Comment:
I think it's a bad idea to start a server on ST's async thread. Many things in ST rely on it and blocking it (potentially for 5 min), will likely cause issues in some cases.

I don't that in theory the user should log in right away and the server should close but things can go wrong very easily and it will be confusing to see a lot of ST not working properly.

We should start a dedicated thread for the server.